### ISSUE TITLE
FAQ: In what situations would this approach be preferable to TFCO?

### ISSUE THREAD

**Issue by kevinrobinson** [fairlearn/fairlearn#439]
hello!  I was reading the TFCO paper ([Cotter et al. 2018](https://arxiv.org/pdf/1809.04198.pdf)), and it has a great section on past work, including this section:
> Agarwal et al. (2018) recently addressed training classifiers with fairness constraints. Like this work, their proposed algorithm is based on the two-player game perspective. Unlike this paper, they assume a zero-sum game, which works because they also assume oracle solvers for the two players, side-stepping the practical issues of dealing with the non-differentible non-convex indicators in the constraints, which is the focus of our algorithmic and theoretical contributions. Similar to this work, they output a stochastic classifier, but do not provide the sparse m+1 solution that we present in this work. They also consider a deterministic solution, which they produce by searching over a grid of values for λ for the best λ. They noted in their experimental section that the resulting deterministic solution was generally as good as their stochastic solutions on test data for those experiments they tried it on. As they note, a grid-search over λ is less ideal as the number of constraints grows.

To check my understanding, the essential differences is that fairlearn asks you to pick the `n` for the grid search over the values of λ, and then the intent of the dashboard visualizations is to pick a model from that grid search based on the (disparity, accuracy) tradeoff.  In constrast, TFCO allows generating a stochastic classifier that essentially picks from different points in that search, but also allows "shrinking" to a smaller model.  And it supports a wider range of "rate constraints" than the metrics described in `fairlearn`.

I'm sure there's also a range of factors related to the ecosystem and developer experience (eg, TF vs. emphasis on sci-kit compatability), so would be curious to hear folks' thoughts on that as well.  If folks want to comment here, I could afterwards repackage thoughts into a PR for https://github.com/fairlearn/fairlearn/pull/436.  Thanks! 👍

**Comment by romanlutz** [comment]
@MiroDudik  can certainly provide more details, but a few things ahead:

- Let's call the repo/project Fairlearn, not the algorithm. Fairlearn contains other algorithms, not just the one(s) from the original paper.
- The algorithm they refer to from the Agarwal et al. paper is called `ExponentiatedGradient` in this repo, and the "deterministic" one is called `GridSearch`.

So picking `n` is something you do for `GridSearch`, but not for `ExponentiatedGradient`.

The dashboard is not tied to any mitigation techniques. You could use it in a standalone fashion, just analyzing metrics of a model. If you provide data from multiple models it will provide you a comparison view (but you can still drill down into the single model view). None of this requires any sort of mitigation, but some of our examples try to showcase how to use grid search output with the dashboard.

Taking a step back now: there are advantages and disadvantages to every mitigation technique. Fairlearn doesn't necessarily need to push people towards one approach over the other. It would be great if we can provide all tools that people care about, and along with that the educational material to help them decide what makes sense in their particular situation. As mentioned above there are plenty of situations where mitigation may not be required at all. So purely from a project scope standpoint I would say that it would be great to have this as part of the toolbox.

... but when I last tried using it it was rather cumbersome (as you'd expect from research code), and it took months to hear back on the technical difficulties. That said, if there's interest from users I don't see why we couldn't put in the effort to integrate it into Fairlearn. Adding tensorflow as a requirement isn't lightweight at all, but there are certain workarounds like adding it as an extension (similar to what we do with matplotlib right now).
To this day I haven't heard anyone complaining about the lack of mitigation techniques, though, so that's all speculation at this point.

**Comment by kevinrobinson** [comment]
@romanlutz Thanks, this is super helpful! 👍

As you said:

> there are advantages and disadvantages to every mitigation technique... it would be great if we can provide... the educational material to help them decide what makes sense in their particular situation

...this is really what I'm trying to understand better, so thanks!  I'm happy to help package this discussion up into docs if that's helpful after. :)

So first, to check my understanding, `ExponentiatedGradient` code is:

> We proceed iteratively by running a no-regret algorithm for the λ-player, while executing the best response of the Q-player. Following Freund & Schapire (1996), the average play of both players converges to the saddle point. We run the exponentiated gradient algorithm (Kivinen & Warmuth,1997) for the λ-player and terminate as soon as the sub optimality of the average play falls below the pre-specified accuracy ν.

I think I see this all in https://github.com/fairlearn/fairlearn/blob/master/fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py#L153.  So looking at the example in 
the [quickstart guide](https://fairlearn.github.io/quickstart.html#mitigating-disparity) about predicting which people make >$50k, I'm trying to understand what that early termination criteria is.  In technical terms, what is the meaning of `nu` with the default 0.5 `_ACCURACY_MUL` constant in https://github.com/fairlearn/fairlearn/blob/master/fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py#L117?

> self.nu = _ACCURACY_MUL * (h(X) - y_train).abs().std() / np.sqrt(n)

Second, I'm trying to understand when I'd want to use this approach versus the grid search.  For example, in the [credit card example](https://github.com/fairlearn/fairlearn/blob/master/notebooks/Grid%20Search%20for%20Binary%20Classification.ipynb), why does it make more sense to use the `GridSearch` versus just using the `ExponentiatedGradient` method?

It's probably good to cut back down the scope to focusing on the different algorithmic approaches and contexts where they'd make more or less sense (vs other ecosystem/community questions), since I've asked too many questions here. :)

**Comment by romanlutz** [comment]
It's absolutely helpful to document this. Feel free to submit PRs as you see fit! :-)

Not sure if I see a question before "Second", but yes, the exponentiated_gradient.py file has the logic from the paper. You may also have noticed some logic related to a linear program (in the code `LP` or `linprog` depending on where you look) which is not mentioned in the paper. It's an optimization @MiroDudik added.

`GridSearch` vs `ExponentiatedGradient` is an interesting comparison! One could think of `GridSearch` as the naive or brute force version of `ExponentiatedGradient`. That has certain limits, though. If you have a binary sensitive feature, you effectively end up with a 1-dimensional grid (if you want to call that a grid at all...). Searching that is not too hard. Once your sensitive features get more values or you consider intersectionality the grid dimensionality explodes pretty quickly. [Sidenote: as you're aware intersectionality should certainly be on people's radar]
At each grid point we end up having to fit our estimator. Let's say we have a 10-dimensional grid. To cover this adequately we'd need a lot of grid points, or we may end up not finding a very good solution. `ExponentiatedGradient` approaches this different by iteratively and intelligently picking the next solution until we either get below our tolerance or hit the iteration limit. The paper provides certain theoretical guarantees so as long as your iteration limit is high enough you'll find a solution that suits your requirements.

Taking a step back: this and a lot more should be explained in the [user guide](https://fairlearn.github.io/user_guide/mitigation.html#reductions), perhaps with examples to illustrate. It's something we're working towards. As you've noticed the documentation is lacking on various fronts, so I can't quite say when this gets better. If you feel like improving any part don't feel shy to do so. Thanks!

**Comment by MiroDudik** [comment]
I really like [Cotter et al. 2018](http://jmlr.csail.mit.edu/papers/volume20/18-616/18-616.pdf)--I think that their line of work is very closely related to ours and they present some nice improvements.

They present several algorithms:
* The high level idea behind Algorithm 1 is the same as our `fairlearn.reductions.ExponentiatedGradient`--they solve the same basic optimization problem. We have implemented some additional tweaks that improve the performance:
  * We use exponentiated gradient instead of projected (sub)gradient. I did some initial benchmarking when writing our paper, and exponentiated gradient seemed to work better, so that's what we have implemented.
  * We check for the duality gap, which allows us to terminate early, as soon as we solve the optimization problem.
  * We carry out the _shrinkage step_ from their paper in each iteration (we call in `LP_step`). This is computationally very cheap, but practically very important, because it allows us to solve the problem in around 10 iterations rather than something like 100 or more--which is how long it would take without it (or if you only used it for sparsification at the very end as they suggest).

The key assumption that Algorithm 1 and ExponentiatedGradient make is that the underlying supervised learner (aka "oracle" aka "base estimator") can solve some hard problems. That's a similar leap of faith that data scientists make when they run _any_ classification algorithm (0/1 classification is NP-hard). In ExponentiatedGradient, we take some additional heuristical steps ([here](https://github.com/fairlearn/fairlearn/blob/fee71c194e5fbc08d2d17f431d30788752adba5c/fairlearn/reductions/_exponentiated_gradient/_lagrangian.py#L101) and [here](https://github.com/fairlearn/fairlearn/blob/fee71c194e5fbc08d2d17f431d30788752adba5c/fairlearn/reductions/_exponentiated_gradient/_lagrangian.py#L173)) to "encourage" the oracle to solve these hard problems, but sometimes the oracle might fail. In that case, ExponentiatedGradient might not reduce the duality gap as fast as implied by theory. A more typical consequence is that the calculated duality gap is small, but incorrect, and so the optimality guarantee would not hold. I have observed this behavior in ExponentiatedGradient in situations when the fairness constraints are harder to satisfy--e.g., when `eps` is close to zero. This is the failure that Cotter et al. are trying to address.

* Their Algorithms 2, 3 and 4 introduce the idea of proxy-Lagrangian.
  * This can be advantageous if you are wrapping a supervised learner (i.e., "reducing to an oracle") that can solve for arbitrary differentiable objectives--for example, if your are trying to fit a neural net in tensorflow. ExponentiatedGradient can still be applied, but it does not exploit the differentiability structure and so the problems that it is creating for the underlying supervised learner might be harder.
  * Algorithms 2-4 do not reduce to supervised learning, but to _optimization_. This means that you cannot for example easily specify them as "wrappers" (or meta-estimators in the sklearn terminology). Also, I don't think they would easily work with classifiers that are not based on optimization--like decision trees or tree ensembles.
  * Algorithm 2 still creates NP-hard subproblems (that are passed to the oracle), but because the problems it creates are differentiable, they might be easier than those created by ExponentiatedGradient. The price that it pays is that its optimality guarantees are weaker than those enjoyed by Algorithm 1 and ExponentiatedGradient. Specifically: note that Eq. 16 has _g&#771;<sub>i</sub>_ in the `inf` whereas the matching expression in Theorem 2 has _g<sub>i</sub>_. That said--it's not clear to me that this distinction matters in practice.
  * Algorithm 3 does not create hard problems, but its scope is more limited, because it assumes that the surrogates are convex--in that case Eq. 16 might bite quite unpleasantly. You could possibly run this algorithm without convexity and it might work in practice, but guarantees don't hold.
  * Algorithm 4 is a heuristic without guarantees, but it might work reasonably well in practice.

So--what's the practitioner to do?

* If the underlying supervised oracle (base estimator) is not optimization based--e.g., when it is fitting a decision tree or an ensemble of trees--then it is not really possible to run Algorithms 2-4.

* If you have access to an optimizer--e.g., you are training something that can be written as a (differentiable) neural net--it makes sense to try one of Algorithms 2-4. In particular, Algorithm 4 seems to do well empirically (although it does not come with guarantees). Besides creating differentiable subproblems it has an additional advantage over ExponentiatedGradient: it does not require solving the subproblems to full precision, and so it can be much faster. But that also opens a can of worms--since you need to tune the learning rate, mini-batch size etc.. So I would still suggest running ExponentiatedGradient at least once (using the hyperparameters that work for the unconstrained version of the problem). And only then embark on Algorithm 4.

**Comment by kevinrobinson** [comment]
@MiroDudik This is amazingly helpful, thank you! 😄  I really appreciate how you're referring to specific reference points in the papers, and pointing out where the code in fairlearn has additional improvements.  I also am thankful for the bit translating this into a practitioner perspective too :)

One quick reaction I have is that there is a lot of complexity to understand how the theoretical work in any of these approaches will apply to any specific problem or scenario. :) So while the 2018 paper says things like:

> Our reductions work for any classifier representation, encompass many definitions of fairness, **satisfy provable guarantees**, and work well in practice.

I think I might have interpreted what terms like 'constraints' and 'guarantees' meant a bit too naively or rigidly, and so this is super helpful in realizing I need to understand this more deeply. 👍 

I'll dig into reading through this more to make sure I'm fully understanding, and although it might take some more questions and clarifications, I'd love to help translate this discussion into user-facing documentation or an example notebook illustrating what you're describing in a particular scenario.
